### ISSUE TITLE
Segfault/Mutex Error

### ISSUE THREAD

**Issue by muchanem** [google/tensorstore#179]
I'm encountering a segmentation fault with the following snippet (the idea of this pattern being that after the first blocking call to result I can pass data to GPU/my model while the asynchronous read starts and I can call the blocking/sync read when the model needs data again)

```python
import tensorstore as ts
import numpy as np

# shape = (1003356160, 2)
idxs = np.load("/net/projects/interp/shuffled_idxs.npy")

dataset = ts.open(
    {
    'driver': 'zarr3',
    'cache_pool': {'total_bytes_limit': 1024*1024*100},
    'recheck_cached_data': 'open',
    'kvstore': {
        'driver': 'file',
        'file_io_concurrency': {'limit': 1024},
        'path': '/net/projects/interp/ts_acts',
        },
    },
    dtype=ts.float32,
    chunk_layout=ts.ChunkLayout(
    read_chunk_shape=[1, 1, 3072],
    write_chunk_shape=[2560, 256, 3072],
    ),
    shape=[3919360, 256, 3072],
).result()

batch_size = 4096
to_load = idxs[0:batch_size*100]
outer, inner = to_load[:,0], to_load[:,1]
next_table = dataset[outer,inner].read()

curr_table = next_table.result()

start = batch_size*100
to_load = idxs[start:start+batch_size*100]
outer, inner = to_load[:,0], to_load[:,1]
next_table = dataset[outer,inner].read()

# model computation here

curr_table = next_table.result()

# model computation here
```

However, on the second blocking sync the code exits with a segmentation fault or a mutex error
```
[mutex.cc : 452] RAW: Lock blocking 0x7f56dc397940   @ 0x7f5aaad70ec5 0x7f5aa98c9e8c 0x7f5aaa622c39 0x7f5aaa6237ac 0x7f5aaa62403f 0x7f5aaa62479a 0x7f5aaa620fb5 0x7f5aaa621cf4 0x7f5aaa621fe4 0x7f5aaa3e295c 0x7f5aaa3e2c3b 0x7f5aaa61ce33 0x7f5aaa61f427 0x7f5aaad9b050
[mutex.cc : 1957] RAW: Check (v & (kMuWait | kMuWrWait)) != kMuWrWait failed: Lock: Mutex corrupt: waiting writer with no waiters: 0x7f56dc17e1f
```
or a `double free or corruption` error. Is there something wrong with this pattern?

**Comment by muchanem** [comment]
This is somehow related to the file_io_concurrency, reducing to 512 from 1024 resolves

**Comment by jbms** [comment]
There shouldn't be a crash regardless of the `file_io_concurrency` setting --- we'll take a look at this.

**Comment by muchanem** [comment]
Yep--just tested with a slightly more complicated version of this pattern and it still crashed with the reduced file_io_concurrency (and even with removing the setting). 

```python
def tensorstore_data_generator(dataset, idxs, batch_size):
    """
    Generate batches of data from a tensorstore database.

    Args:
        dataset (ts.TensorStore): Full dataset
        batch_size (int): Size of each batch.
        idxs: List of indices to slice from the dataset.

    Yields:
        jax.Array: A batch of data.
    """
    to_load = idxs[0:batch_size*100]
    outer, inner = to_load[:,0], to_load[:,1]
    next_table = dataset[outer,inner].read()
    for start in range(batch_size*100, len(idxs), batch_size*100):
        curr_table = next_table.result()
        to_load = idxs[start:min(start+batch_size*100, idxs.shape[0])]
        outer, inner = to_load[:,0], to_load[:,1]
        next_table = dataset[outer,inner].read()
        for i in range(100):
            yield jnp.array(curr_table[i*batch_size:min((i+1)*batch_size, len(curr_table))])
```

Here's the snippet that crashes regardless of the file_io_concurrency setting.

**Comment by laramiel** [comment]
The index file is, I assume, a variant from the other issue, and something like:

```
  # shape = (1003356160, 2)
  if os.path.isfile('/tmp/shuffled_idxs.npy'):
    idxs = np.load("/tmp/shuffled_idxs.npy")
  else:
    idxs = np.array(list(product(list(range(2304000)),list(range(256)))))
    shuffle(idxs)
    np.random.shuffle(idxs)
    np.save('/tmp/shuffled_idxs.npy', idxs)
```

**Comment by muchanem** [comment]
Yes, a variant from the other issue. The snippet below is how I generated the indices that are being loaded from disk. The idea being that I want a random shuffling of the entire dataset over its first two dimensions
```python
x = np.arange(3919360)
y = np.arange(256)
idxs = np.dstack(np.meshgrid(x, y)).reshape(-1, 2)
rng = np.random.default_rng()

rng.shuffle(idxs)
```

**Comment by laramiel** [comment]
Just FWIW: It may make sense to try a smaller write chunk shape. The current shape writes 8GB chunk files; using  `[256, 256, 3072]` writes 800 MB chunk files, and `[64, 256, 3072]` writes 100MB chunk files.

**Comment by muchanem** [comment]
I'm using sharded chunking right now (as I understand it, setting both a read and write chunk size implicitly uses sharding?)--would there be any benefit to have a smaller written chunk shape?

**Comment by jbms** [comment]
Can you confirm which version of tensorstore you are using?

**Comment by jbms** [comment]
Also if you are able to produce a self-contained script that reproduces the crash, ideally just on the local filesystem, that would be very helpful.

**Comment by muchanem** [comment]
This was on tensorstore `v0.1.56`, after upgrading to the latest version (and also without using sharded chunking) the issue has gone away. I will update if it returns

**Comment by jbms** [comment]
Thanks, this may have been fixed by https://github.com/google/tensorstore/commit/f021374867642099a17f3b6f8cf5a65d06f0cbfa

**Comment by laramiel** [comment]
Ok, fwiw, I was trying to repro with something like this:

```
import numpy as np
import tensorstore as ts

BATCH_SIZE = 4096 * 100
SHAPE = (391936, 256, 3072)
PATH = '/mnt/disks/slow/issue_179'

def _open_tensorstore(**kwargs):
  dataset = ts.open(
      {
          'driver': 'zarr3',
          'cache_pool': {'total_bytes_limit': 1024 * 1024 * 100},
          'recheck_cached_data': 'open',
          'kvstore': {
              'driver': 'file',
              'file_io_concurrency': {'limit': 1024},
              'path': PATH,
          },
      },
      **kwargs,
      dtype=ts.float32,
      chunk_layout=ts.ChunkLayout(
          read_chunk_shape=[1, 1, 3072],
          write_chunk_shape=[256, 256, 3072],
      ),
  ).result()
  if (
      dataset.shape[0] != SHAPE[0]
      or dataset.shape[1] != SHAPE[1]
      or dataset.shape[2] != SHAPE[2]
  ):
    dataset = dataset.resize(exclusive_max=SHAPE, expand_only=True).result()
  return dataset


def _random_idx(shape, n):
  return np.array([
      [np.random.randint(0, shape[0]), np.random.randint(0, shape[1])]
      for _ in range(n)
  ])


def fill_tensorstore(dataset):
  # This is quite large and is not suitable for a unit test.
  write_shape = dataset.chunk_layout.write_chunk.shape
  data = np.float32(np.random.rand(*write_shape))
  awaitables = []

  print('Fill tensorstore')
  if dataset.storage_statistics(query_fully_stored=True).result().fully_stored:
    print('... Fully stored')
    return

  def _wait_all():
    nonlocal awaitables
    for x in awaitables or []:
      x.result()
    awaitables = []

  def _write_shard(stats, shard):
    if not stats.not_stored:
      return
    nonlocal awaitables
    nonlocal data
    awaitables.append(shard.write(data))
    if len(awaitables) > 10:
      _wait_all()

  for s0 in range(0, dataset.shape[0], write_shape[0]):
    for s1 in range(0, dataset.shape[1], write_shape[1]):
      shard = dataset[s0 : s0 + write_shape[0], s1 : s1 + write_shape[1]]
      f = shard.storage_statistics(query_not_stored=True)
      f.add_done_callback(
          lambda x, shard=shard: _write_shard(x.result(), shard)
      )
      awaitables.append(f)
      if len(awaitables) > 10:
        _wait_all()

  _wait_all()
  print('End writes')


def test_issue_179(dataset):
  # shape = (1003356160, 2)

  print('Beginning test')

  to_load = _random_idx(dataset.shape, BATCH_SIZE)
  outer, inner = to_load[:, 0], to_load[:, 1]
  next_table = dataset[outer, inner].read()

  curr_table = next_table.result()

  to_load = _random_idx(dataset.shape, BATCH_SIZE)
  outer, inner = to_load[:, 0], to_load[:, 1]
  next_table = dataset[outer, inner].read()

  # model computation here

  curr_table = next_table.result()

  # model computation here


def main():
  print('Opening ts')
  fill_tensorstore(_open_tensorstore())

  while True:
    test_issue_179(_open_tensorstore())
    print(ts.experimental_collect_matching_metrics('/tensorstore'))


main()
```

**Comment by laramiel** [comment]
I was also using a delay device mapper to slow down local disk performance:

```
 # Create a block device with a 200ms delay  (the device is on a GCE vm)
 DEV=/dev/disk/by-id/google-disk-delaytest
 BLOCKS=$(sudo blockdev --getsize $DEV)

 sudo mkfs.ext4 -m 0 -E lazy_itable_init=0,lazy_journal_init=0,discard $DEV
 echo "0 $BLOCKS delay $DEV 0 200" | sudo dmsetup create dm-slow

 sudo mkdir /mnt/disks/slow
 sudo mount -o discard,defaults /dev/mapper/dm-slow /mnt/disks/slow
 sudo chmod a+rwX /mnt/disks/slow

 # Use /mnt/disks/slow in tensorstore.
 ```
